client.go

```diff
@@ -0,0 +1,331 @@
+package client
+
+import (
+	"bytes"
+	"encoding/json"
+	"fmt"
+	"io"
+	"net/http"
+	"strconv"
+	"time"
+)
+
+// tokenExpiredError Error
+type tokenExpiredError struct {
+}
+
+func (e tokenExpiredError) Error() string {
+	return "Expired Token"
+}
+
+// NotFound Error
+type NotFound struct {
+}
+
+func (e NotFound) Error() string {
+	return "Token Not Found"
+}
+
+// VisualizationError errors for Visualization client
+type VisualizationError struct {
+	code        string
+	message     string
+	description string
+}
+
+// Error generate a error message.
+// If Code is zero, we know it's not a http error.
+func (e VisualizationError) Error() string {
+	return fmt.Sprintf("ERROR: %s", e.description)
+}
+
+// VisualizationClient client for Visualization
+type VisualizationClient struct {
+	url            string
+	client         *http.Client
+	token          AuthToken
+	openstackToken string
+}
+
+// NewVisualizationClient returns client with token
+func NewVisualizationClient(url string, token AuthToken) (*VisualizationClient, error) {
+	client := http.Client{}
+	return &VisualizationClient{client: &client, url: url, token: token}, nil
+
+}
+
+// A JSONMessage contains the json error message received when http request failed
+type JSONMessage struct {
+	Message string `json:"message"`
+}
+
+func (v *VisualizationClient) reIssue() bool {
+	_, err := v.OpenstackAuth()
+	return err == nil
+}
+
+// authorizedHttpRequest for token checking and reauth
+func (v *VisualizationClient) authorizedHttpRequest() (err error) {
+	// validate token
+	if v.token == (AuthToken{}) {
+		return NotFound{}
+	}
+	tokenExpires := v.token.Token.ExpiresAt.UnixNano() / 1000000
+	now := time.Now().UnixNano() / 1000000
+	if tokenExpires < now {
+		if v.reIssue() {
+			return v.authorizedHttpRequest()
+		}
+	}
+	return err
+}
+
+// httpRequest handle the request to Grafana server.
+//It returns the response body and a error if something went wrong
+func (v *VisualizationClient) httpRequest(method string, url string, body io.Reader) (result io.Reader, err error) {
+	err = v.authorizedHttpRequest()
+	if err != nil {
+		return
+	}
+	request, err := http.NewRequest(method, url, body)
+	request.Header.Set("Content-Type", "application/json")
+	request.Header.Add("X-OpenStack-Auth-Token", v.openstackToken)
+	bearer := fmt.Sprintf("Bearer %v", v.token)
+	request.Header.Add("Authorization", bearer)
+
+	response, err := v.client.Do(request)
+	if err != nil {
+		return result, VisualizationError{"0", "Unable to perform the http request", "request failed"}
+	}
+
+	//	defer response.Body.Close()
+	if response.StatusCode != 200 {
+		dec := json.NewDecoder(response.Body)
+		var Mess JSONMessage
```

Comment:
If status code is not 200, the response is Error model itself, so just decode VisualizationError model